### DISCUSSION TITLE
How to dynamically set visual (RelativePanel) AttachedProperties of Items in ItemsControl

### DISCUSSION THREAD

**Discussion by NoizeDaemon** [AvaloniaUI/Avalonia#10394] [Q&A] [answered]
So I started using Avalonia like a week ago, and all I did was trying to get the `RelativePanel` to behave like I want with a minimum of XAML/(Multi)bindings/Converters required. That did work out somewhat well using `AttachedProperties`.. only to realise afterwards that it does not work when used in an `ItemsControl` 😂

The Bindings show up correctly in the Debugger, but they do not change the visuals - the Items all stay in the upper left corner.
Even after forcing the panel to refresh with `InvalidateMeasure`/`InvalidateArrange`.

![image](https://user-images.githubusercontent.com/42150677/219903858-ece82a98-ecf5-4464-8cdf-82723bb3dd2b.png)

As I understand it, the `DataTemplate` is not meant to be used for Binding Layout/Visual Properties like that.
In addition to that there's a `ContentPresenter` between my Items and the `RelativePanel`, so that the properties like `RelativePanel.Below` can't "find" a parent RelativePanel to become effective...

I know there's `Styles`, but AFAIK I can only bind to Static Ressources with them, correct? Kinda need them to be dynamic. Also I would not know how to change them on a per item basis.

**My question is simply how to make Bindings inside of an `ItemsControl` work with the `RelativePanel` (or `Canvas` would be fine too). Or how to work sensibly around that.**

Is there a common way of doing this? Or can someone help me figure out a way?

I had a few ideas, maybe they're not super stupid & someone smarter could make them work or give me a hint:

- Avoid the ContentPresenter:
While going through the API reference I stumbled onto something like "ItemIsItsOwnContainer", but I can't find it anymore.
What Controls don't require a `ContentPresenter` inbetween the Panel & the Item?
Is there a way to force a Control like a Grid to be it's own Container?
- Inherit PropertyValues "upwards"/Bind to `ContentPresenter` directly
I experimented with setting the desired properties in the `ContentPresenter` itself and with setting the PropertyValues for the parent when the Bindings for my Items update. That gets me this:
`RelativePanel error: Element does not exist in the current context: RightOf`
That's probably related to the RelativePanel showing up in the Visualtree but not in the LogicalTree. I was wondering about that...
- Using AvaloniaBehaviours (somehow)..
That sadly isn't an option right now, since I'm on `11.0.999-cibuild0030108-beta`, which fixes Bindings with the RelativePanel in the first place. AFAIK that version does not work with Behaviours yet (but I could try again).
- Listening to changes in my SourceList and "manually" adding the Items to a RelativePanel
That would probably work, but makes it a nightmare to handle changes dynamically - that's kinda what the ItemsControl is for...

Any ideas are welcome :)

Here's the (not so minimal anymore) sample project: https://github.com/NoizeDaemon/MinimalSample
There's different tests concerning the RelativePanel in there, but they are labeled.. Can remove them though if needed/if it makes everything too cluttered.

But here's also some of the relevant code - there's a lot of weird things happening because at this point I was just trying everything that sounded vaguely like sth that could help 😂

Model:
(`IsPrime` does nothing in this project, the actual project is creating a dynamic Ulam Spiral, that's why this is still here 😅)
```
namespace MinimalSample.Models
{
    public partial class NumberItem : ObservableObject
    {
        [ObservableProperty]
        private string name;

        [ObservableProperty]
        private int number;

        [ObservableProperty]
        private bool isPrime;

        [ObservableProperty]
        private Direction? direction;

        [ObservableProperty]
        private string? neighbor;
    }

    public enum Direction
    {
        Unset,
        RightOf,
        Above,
        LeftOf,
        Below
    }
}
```

ViewModel:
```
...
        private readonly SourceList<NumberItem> numberItemsSourceList = new();
        private readonly ReadOnlyObservableCollection<NumberItem> numberItems;
        public ReadOnlyObservableCollection<NumberItem> NumberItems => numberItems;

        public MainViewModel()
        {
            numberItemsSourceList
                .Connect()
                .ObserveOn(RxApp.MainThreadScheduler)
                .Bind(out numberItems)
                .Subscribe();


            numberItemsSourceList.Add(new NumberItem
            {
                Name = "N1",
                Number = 1,
                IsPrime = false,
                Direction = Direction.Unset
            });
            ....
        }

        [RelayCommand]
        void AddDirection(string directionInput)
        {
            var newDirection = Enum.Parse<Direction>(directionInput);
            numberItemsSourceList.Add(new NumberItem
            {
                Name = "N" + (numberItemsSourceList.Count + 1),
                Number = numberItemsSourceList.Count + 1,
                IsPrime = false,
                Direction = newDirection,
                Neighbor = "N" + numberItemsSourceList.Count
            });
        }
...
```

MainView (XAML):
```
...
<!--In ItemsControl with RelativePanel as ItemsPanelTemplate-->
<StackPanel Grid.Row="2" Grid.ColumnSpan="3" Orientation="Horizontal" HorizontalAlignment="Center" Spacing="50" >
	<Button Command="{Binding AddDirectionCommand}" CommandParameter="LeftOf" IsEnabled="{Binding AddLeftEnabled}">Add Left</Button>
	<Button Command="{Binding AddDirectionCommand}" CommandParameter="RightOf" IsEnabled="{Binding AddRightEnabled}">Add Right</Button>
	<Button Command="{Binding AddDirectionCommand}" CommandParameter="Above" IsEnabled="{Binding AddAboveEnabled}">Add Above</Button>
	<Button Command="{Binding AddDirectionCommand}" CommandParameter="Below" IsEnabled="{Binding AddBelowEnabled}">Add Below</Button>
	<Button Click="OnResetButtonClick">Invalidate Measure/Arrange</Button>
</StackPanel>

<!--Method 1b-->
<DockPanel Grid.Row="3" Grid.Column="0">
	<TextBlock Text="Method 1b" TextAlignment="Center" DockPanel.Dock="Top"/>
	<Border BorderBrush="Cyan" BorderThickness="2" DockPanel.Dock="Bottom">
		<ItemsControl x:Name="itemsControl1b" Items="{Binding NumberItems}">
			<ItemsControl.ItemsPanel>
				<ItemsPanelTemplate>
					<RelativePanel Name="relativePanel1b"/>
				</ItemsPanelTemplate>
			</ItemsControl.ItemsPanel>
			<ItemsControl.ItemTemplate>
				<DataTemplate>
					<Grid x:Name="{Binding Name}"
						  local:EasyAlignWith.Direction="{Binding Direction}">
						<local:EasyAlignWith.Target>
							<MultiBinding Converter="{StaticResource stringToControlConverter}">
								<Binding Path="$parent"/>
								<Binding Path="Neighbor"/>
							</MultiBinding>
						</local:EasyAlignWith.Target>
						<Rectangle Width="25" Height="25" Fill="White"/>
						<TextBlock Text="{Binding Number}" Foreground="Black" HorizontalAlignment="Center" VerticalAlignment="Center"/>
					</Grid>
				</DataTemplate>
			</ItemsControl.ItemTemplate>
		</ItemsControl>
	</Border>
</DockPanel>
...
```

MainView (CS):
```
...
        //For Method 1b:
        void OnResetButtonClick(object sender, RoutedEventArgs e)
        {
            var visuals = this.GetVisualDescendants();

            //Reset RelativePanels - does nothing
            var panels = visuals.Where(x => x is RelativePanel);

            foreach(var panel in panels)
            {
                ((RelativePanel)panel).InvalidateMeasure();
                ((RelativePanel)panel).InvalidateArrange();
            }
        }
...
```

Converter:
```
...
        public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
        {
            if (values[0] == AvaloniaProperty.UnsetValue ||
                values[1] == AvaloniaProperty.UnsetValue ||
                values[1] is null)
            {
                return BindingOperations.DoNothing;
            }
            else if ((string)values[1] == "")
            {
                return AvaloniaProperty.UnsetValue;
            }
            else
            {
                Control parent = (Control)values[0];
                string target = (string)values[1];

                if (parent is ItemsControl)
                {
                    var children = ((ItemsControl)parent).GetVisualDescendants();
                    Control control = (Control)children.Where(x => x.Name == target).First();
                    return control ?? BindingOperations.DoNothing;
                }
                else
                {
                    Control control = parent.FindControl<Control>(target);
                    return control ?? BindingOperations.DoNothing;
                }
            }
        }
...
```

AttachedProperties:
(They work fine outside of ItemsControl, but if anyone wondered what the `EasyAlignWith` that I am binding to does)
```
    public class EasyAlignWith : RelativePanel
    {
        static EasyAlignWith()
        {
            AffectsParentArrange<RelativePanel>(DirectionProperty, TargetProperty);
            AffectsParentMeasure<RelativePanel>(DirectionProperty, TargetProperty);

            DirectionProperty.Changed.AddClassHandler<AvaloniaObject>(OnDirectionTargetPropertyChanged);
            TargetProperty.Changed.AddClassHandler<AvaloniaObject>(OnDirectionTargetPropertyChanged);
        }

        private static void OnDirectionTargetPropertyChanged(AvaloniaObject obj, AvaloniaPropertyChangedEventArgs e)
        {
            Direction direction = obj.GetValue(DirectionProperty);
            var targetValue = obj.GetValue(TargetProperty);

            if (direction != Direction.Unset && targetValue != AvaloniaProperty.UnsetValue && targetValue != null)
            {
                if (direction is Direction.RightOf)
                {
                    if (obj.GetValue(LeftOfProperty) == targetValue) RelativePanel.SetLeftOf(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(AboveProperty) == targetValue) RelativePanel.SetAbove(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(BelowProperty) == targetValue) RelativePanel.SetBelow(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(AlignHorizontalCenterWithProperty) == targetValue) RelativePanel.SetAlignHorizontalCenterWith(obj, AvaloniaProperty.UnsetValue);

                    RelativePanel.SetAlignVerticalCenterWith(obj, targetValue);
                    RelativePanel.SetRightOf(obj, targetValue);
                }
                else if (direction is Direction.LeftOf)
                {
                    if (obj.GetValue(RightOfProperty) == targetValue) RelativePanel.SetRightOf(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(AboveProperty) == targetValue) RelativePanel.SetAbove(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(BelowProperty) == targetValue) RelativePanel.SetBelow(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(AlignHorizontalCenterWithProperty) == targetValue) RelativePanel.SetAlignHorizontalCenterWith(obj, AvaloniaProperty.UnsetValue);

                    RelativePanel.SetAlignVerticalCenterWith(obj, targetValue);
                    RelativePanel.SetLeftOf(obj, targetValue);
                }
                else if (direction is Direction.Above)
                {
                    if (obj.GetValue(RightOfProperty) == targetValue) RelativePanel.SetRightOf(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(LeftOfProperty) == targetValue) RelativePanel.SetLeftOf(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(BelowProperty) == targetValue) RelativePanel.SetBelow(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(AlignVerticalCenterWithProperty) == targetValue) RelativePanel.SetAlignVerticalCenterWith(obj, AvaloniaProperty.UnsetValue);


                    RelativePanel.SetAlignHorizontalCenterWith(obj, targetValue);
                    RelativePanel.SetAbove(obj, targetValue);
                }
                else if (direction is Direction.Below)
                {
                    if (obj.GetValue(RightOfProperty) == targetValue) RelativePanel.SetRightOf(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(LeftOfProperty) == targetValue) RelativePanel.SetLeftOf(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(AboveProperty) == targetValue) RelativePanel.SetAbove(obj, AvaloniaProperty.UnsetValue);
                    if (obj.GetValue(AlignVerticalCenterWithProperty) == targetValue) RelativePanel.SetAlignVerticalCenterWith(obj, AvaloniaProperty.UnsetValue);


                    RelativePanel.SetAlignHorizontalCenterWith(obj, targetValue);
                    RelativePanel.SetBelow(obj, targetValue);
                }
            }
        }
```

For anyone who's wondering why I want to do this:
I want to arrange my NumberItems in a spiral pattern. Like so:
5 4 3
6 1 2
7 8 9 etc.
& dynamically change/recolour/rearrange them.

I could do this in different ways as well, like with a Canvas or (Data)Grid.
I just started with the RelativePanel because it sounded the simplest - boy, was I wrong 😂
At this point it is more about seeing where the limits lie, trying to make sth work.

So please no suggestions about how to achieve the spiral thingy without an ItemsControl.
Figuring that out is half the fun of coding ❤ But since I already spent a week on the RelativePanel thing, the fun is slowly starting to diminish.. 😅

**Comment by NoizeDaemon** [comment]
I definitely made it work in code-behind haha 😂
Still not dynamic/bound to the SourceList, but I can at least play Snake now 🤣

### So the questions above still stand ^^^
but at least I had a good laugh 😁

![image](https://user-images.githubusercontent.com/42150677/219909577-9e797830-275c-43b0-8578-823788befc66.png)

**Comment by jp2masa** [comment]
I think you are looking for one of these:

- https://github.com/AvaloniaUI/Avalonia/issues/3100#issuecomment-753452752
- https://github.com/AvaloniaUI/Avalonia/discussions/10018#discussioncomment-4726892

**Reply by NoizeDaemon** [reply]
I could not make `ItemContainerStyle` work for my specific scenario, but it's useful info for the future! 

Thank you! 😊

**Comment by ahopper** [comment, marked as answer]
You can bind the ContentPresenter to properties like this
https://github.com/kekekeks/Avalonia.BattleCity/blob/2735e216327636d5970b40c9ccf9141d586bb1bc/MainWindow.xaml#L9-L20

**Reply by NoizeDaemon** [reply]
That looks pretty much like what I need.

How is it that your Style Bindings bind to the appropriate properties (like `Location`) on the `GameObject` in your `Items="{Binding GameObjects}"`? Like the binding works on an Item level, correct?

If I try to do that like this:
```
...
xmlns:model="clr-namespace:MinimalSample.Models;assembly=MinimalSample"
...
	<UserControl.Styles>
		<Style Selector="#itemsControl1b > ContentPresenter">
			<Setter Property="RelativePanel.AlignHorizontalCenterWithPanel" Value="{Binding Neighbor, Converter={x:Static StringConverters.IsNullOrEmpty}}"/>
			<Setter Property="RelativePanel.AlignVerticalCenterWithPanel" Value="{Binding Neighbor, Converter={x:Static StringConverters.IsNullOrEmpty}}"/>
		</Style>
	</UserControl.Styles>
...
				<ItemsControl x:Name="itemsControl1b" Items="{Binding NumberItems}">
					<ItemsControl.ItemsPanel>
						<ItemsPanelTemplate>
							<RelativePanel Name="relativePanel1b"/>
						</ItemsPanelTemplate>
					</ItemsControl.ItemsPanel>
					<ItemsControl.ItemTemplate>
						<DataTemplate DataType="{x:Type model:NumberItem}">
							<Grid x:Name="{Binding Name}"
...
							</Grid>
						</DataTemplate>
					</ItemsControl.ItemTemplate>
				</ItemsControl>
```
There's an error telling me it can't find `Neighbor` in my `MainViewModel` - well, ofc not since it's part of the Model that the ItemsControl is binding to a list of to.

So do I need to set a different DataContext for the Style or sth similar? But that's not what you did here. But then again, maybe I am missing sth.

Edit:
Just realised that is someone else's project.
Is it okay to link/mention the creator?

**Reply by NoizeDaemon** [reply]
YES, GOT IT <3

I am using CompiledBindings, so instead I had to do this:
`<Setter Property="RelativePanel.AlignHorizontalCenterWithPanel" Value="{ReflectionBinding Neighbor, Converter={x:Static StringConverters.IsNullOrEmpty}}"/>`

Hell yeah! <3 Thank you!

**Reply by timunie** [reply]
ReflectionBinding is not needed if you specify x:DataType or you can use casting similar to how it's done in C#

**Reply by NoizeDaemon** [reply]
Oh good to know! Thanks!
